### PR TITLE
upgrading TF to be able to use Bellman

### PR DESCRIPTION
**Pull request by hstojic**
packages related to RL agents had to be bumped up

2 unresolved errors related to RL code, @IanRDavies it would be great if you could chip in with these, since you have written this code you would be able to fix it far quicker than I can

1. first is related to the class `OneHotCategoricalProjectionNetwork`, here is the error output:
```python
_________________ test_one_hot_categorical_projection_network __________________

    def test_one_hot_categorical_projection_network():
        """
        Test the networks used as action heads.
        This tests initialisation and the forward pass.
        """
        # Set up for a single action head with 5 actions in the subspace.
        num_actions = 5
        sample_spec = BoundedTensorSpec(
            shape=tf.TensorShape((num_actions,)),
            dtype=tf.dtypes.float32,
            minimum=0,
            maximum=1,
            name="action"
        )
        action_head = OneHotCategoricalProjectionNetwork(sample_spec, num_actions)
        # Test the initialisation.
        assert hasattr(action_head, "_projection_layer") and action_head._projection_layer is not None
        assert hasattr(action_head, "_output_spec") and isinstance(action_head._output_spec,
                                                                   DistributionSpec)
        # Test the forward pass (assuming a final output of the shared layers of dimension 64).
        inputs = tf.convert_to_tensor(np.random.random((1, 100, 64)))
        num_batch_dims = 2
>       action_dist = action_head(inputs, num_batch_dims)

test_multi_headed_softmax_network.py:173: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/tf_agents/networks/network.py:415: in __call__
    outputs, new_state = super(Network, self).__call__(**normalized_kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tfp.distributions.OneHotCategorical 'OneHotCategoricalProjectionNetwork_OneHotCategorical' batch_shape=[1, 100, 1] event_shape=[5] dtype=float32>

    def __iter__(self):
>     raise TypeError('{!r} object is not iterable'.format(type(self).__name__))
E     TypeError: 'OneHotCategorical' object is not iterable

/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/tensorflow_probability/python/distributions/distribution.py:754: TypeError
```

2. here its the `ParallelPyEnvironment` in TF-agents that is causing issues, here is the output for one instance
```python
__________________________________________________ test_rl_env_from_snc_env_action_space_dims_multiple_resource_sets ___________________________________________________

    def test_rl_env_from_snc_env_action_space_dims_multiple_resource_sets():
        """
        Tests the formation and stability of the action space dimensions of the environment through
        the RL environment pipeline in a more complex setting.
        """
        # Set the environment name for this case as the asserts are difficult to make as variables.
        env_name = 'double_reentrant_line_shared_res_homogeneous_cost'
        # Set up the environment parameters.
        # Environment parameters do not affect the test result here.
        env = load_scenario(env_name, job_gen_seed=10).env
        rl_env, action_space_dims = rl_env_from_snc_env(env, discount_factor=0.99, for_tf_agent=False)
>       _, action_space_dims_tf = rl_env_from_snc_env(env, discount_factor=0.99, for_tf_agent=True)

test_rl_environment_wrapper.py:795: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../src/snc/environments/rl_environment_wrapper.py:621: in rl_env_from_snc_env
    parallel_env = TFPyEnvironment(ParallelPyEnvironment(env_constructors))
/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/gin/config.py:1069: in gin_wrapper
    utils.augment_exception_message_and_reraise(e, err_str)
/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/gin/utils.py:41: in augment_exception_message_and_reraise
    raise proxy.with_traceback(exception.__traceback__) from None
/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/gin/config.py:1046: in gin_wrapper
    return fn(*new_args, **new_kwargs)
/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/tf_agents/environments/parallel_py_environment.py:90: in __init__
    self.start()
/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/tf_agents/environments/parallel_py_environment.py:104: in start
    env.start(wait_to_start=self._start_serially)
/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/tf_agents/environments/parallel_py_environment.py:270: in start
    mp_context = multiprocessing.get_context()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

method = None

    def get_context(method: Text = None) -> _multiprocessing.context.BaseContext:
      """Get a context: an object with the same API as multiprocessing module.
    
      Args:
        method: (Optional.) The method name; a Google-safe default is provided.
    
      Returns:
        A multiprocessing context.
    
      Raises:
        RuntimeError: If main() was not executed via handle_main().
      """
      if not multiprocessing_core.initialized():
>       raise RuntimeError(_NOT_INITIALIZED_ERROR)
E       RuntimeError: Unable to load multiprocessing context.
E       
E       Please ensure that you properly initialize your program by wrapping your main()
E       call:
E       
E       def main(argv):
E         ...
E       
E       if __name__ == '__main__':
E         tf_agents.system.multiprocessing.handle_main(main, extra_state_savers=...)
E       
E       or, if using absl.app:
E       
E       if __name__ == '__main__':
E         tf_agents.system.multiprocessing.handle_main(
E             functools.partial(absl.app.run, main), extra_state_savers=...)
E       
E       
E       For unit tests, this also means wrapping your test.main using handle_test_main:
E       
E       if __name__ == '__main__':
E         tf_agents.system.multiprocessing.handle_test_main(
E             tf.test.main, extra_state_savers=...)
E       
E       or
E       
E       if __name__ == '__main__':
E         tf_agents.system.multiprocessing.handle_test_main(
E             tf_agents.utils.test_utils.main, extra_state_savers=...)
E       
E       If you are in interactive mode (e.g. python console, ipython, jupyter notebook)
E       use:
E       
E       tf_agents.system.multiprocessing.enable_interactive_mode(
E           extra_state_savers=...)
E       
E       For more details on state savers, see the docstrings for
E       `tf_agents.multiprocessing.handle_*` and:
E       
E       https://pythonspeed.com/articles/python-multiprocessing/
E       
E         In call to configurable 'ParallelPyEnvironment' (<class 'tf_agents.environments.parallel_py_environment.ParallelPyEnvironment'>)

/home/hstojic/.cache/pypoetry/virtualenvs/snc-c2EbjbA5-py3.7/lib/python3.7/site-packages/tf_agents/system/system_multiprocessing.py:90: RuntimeError
```

regarding the hedgehog code, package `cvxpy` produces a lot of deprecation  warnings about not using the new version of linear algebra operators, it makes it very difficult to read the test output

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -14,26 +14,26 @@ packages = [
 # Note: Poetry will exclude files declared in .gitignore when creating the python package file.
 
 [tool.poetry.dependencies]
-python = "^3.7"
+python = "<3.8,>=3.7"
```
Comment (hstojic) [inline pyproject.toml:17]: had to be changed since the cplex package does not work with python 3.8

**Comment by hstojic** [comment]
fixed issue no 2)

**Comment by IanRDavies** [comment]
Issue 1 appears to be down to changes in what TF-Agents expects now as compared to what it expected back when the code was originally written. The TF agents [`Network`](https://github.com/tensorflow/agents/blob/29daabcc11b277a914f6b848d44c30b4aabbf659/tf_agents/networks/network.py#L114) is [looking to receive](https://github.com/tensorflow/agents/blob/29daabcc11b277a914f6b848d44c30b4aabbf659/tf_agents/networks/network.py#L392) a set of kwargs it can then pass to a Keras layer's [`__call__`](https://github.com/tensorflow/agents/blob/29daabcc11b277a914f6b848d44c30b4aabbf659/tf_agents/networks/network.py#L415). It tries to unpack the kwargs (using `**kwargs` on the arguments collected by `tensorflow.python.utils.tf_inspect.getcallargs`) where it only gets a `OneHotCategorical` distribution which is not iterable. I believe it expects an output and a policy state. We don't currently have a policy state.

I think we could resolve the issue by ensuring that what our distributions/policies return is in line with what TF Agents expects. However, I am not up-to-date with what TF Agents may expect and I've been burnt by poor documentation before. Do you have a good idea or any experience of what TF agents may expect in this case? I find the use of `tensorflow.python.utils.tf_inspect.getcallargs` quite opaque.

**Comment by hstojic** [comment]
Thanks @IanRDavies  for looking into this. You would still be much faster in fixing this, so do you have time to give it a go?

**Comment by IanRDavies** [comment]
I pushed a partial fix to issue 1 but there is now a TensorSpec mismatch issue that TF Agents has zero documentation on that I can find easily. I remember fixing such issues before but not how. Still one step closer to a solution with several tests now fixed.

**Comment by hstojic** [comment]
Thanks @IanRDavies, your last commit resolved things - all tests pass, @sergiovalmac ready for a review!

**Review by sergiovalmac — COMMENTED** [review]
Great work guys! Just a question regarding setting the seed to a default value and a couple of very minor suggestions.

File: src/snc/environments/rl_environment_wrapper.py [hunk 2]
```diff
@@ -597,7 +598,7 @@ def rl_env_from_snc_env(env: crw.ControlledRandomWalk, discount_factor: float,
     if not for_tf_agent:
         return base_env, action_space_dims
 
-    def env_constructor(rl_env, seed):
+    def env_constructor(rl_env, seed=1):
```
[the remark this answers is not in the document]
Comment (sergiovalmac) [reply src/snc/environments/rl_environment_wrapper.py]: I see. Sounds good!

**Review by sergiovalmac — APPROVED** [review]
Excellent!

**Pull request merged**